ode/unit_test/Test_ODE_Newton.hpp

```diff
@@ -0,0 +1,542 @@
+  }
+  Kokkos::deep_copy(x, x_h);
+
+  Kokkos::RangePolicy<execution_space> my_policy(0, 1);
+  NewtonSolve_wrapper solve_wrapper(mySys, params, x, rhs, update, J, tmp,
+                                    status);
+
+  Kokkos::parallel_for(my_policy, solve_wrapper);
+
+  auto status_h = Kokkos::create_mirror_view(status);
+  Kokkos::deep_copy(status_h, status);
+  EXPECT_TRUE(status_h(0) == newton_solver_status::NLS_SUCCESS);
+
+  Kokkos::deep_copy(x_h, x);
+  Kokkos::deep_copy(r_h, rhs);
+#ifdef HAVE_KOKKOSKERNELS_DEBUG
+  std::cout << "Non-linear problem solution and residual:" << std::endl;
+  std::cout << "  [(";
+  for (int eqIdx = 0; eqIdx < mySys.neqs; ++eqIdx) {
+    std::cout << " " << x_h(eqIdx);
+  }
+  std::cout << " ), " << KokkosBlas::serial_nrm2(rhs) << ", (";
+  for (int eqIdx = 0; eqIdx < mySys.neqs; ++eqIdx) {
+    std::cout << " "
+              << Kokkos::abs(x_h(eqIdx) - solution[eqIdx]) /
+                     Kokkos::abs(solution[eqIdx]);
+  }
+  std::cout << " )]" << std::endl;
+#else
+  (void)solution;
+#endif
+}
+
+// Quadratic equation
+// x^2 - x - 2 = 0
+// Solution: x = 2 or x = -1
+// Derivative 2*x - 1
+template <typename execution_space, typename scalar_type>
+struct QuadraticEquation {
+  using vec_type = Kokkos::View<scalar_type*, execution_space>;
+  using mat_type = Kokkos::View<scalar_type**, execution_space>;
+
+  static constexpr int neqs = 1;
+
+  QuadraticEquation() {}
+
+  KOKKOS_FUNCTION void residual(const vec_type& y, const vec_type& f) const {
+    f(0) = y(0) * y(0) - y(0) - 2;
+  }
+
+  KOKKOS_FUNCTION void jacobian(const vec_type& y, const mat_type& jac) const {
+    jac(0, 0) = 2 * y(0) - 1;
+  }
+};
+
+// Trigonometric equation
+// f(x) = cos(x) - x = 0
+// Solution: 0.739085
+// f'(x) = -sin(x) - 1
+template <typename execution_space, typename scalar_type>
+struct TrigonometricEquation {
+  using vec_type = Kokkos::View<scalar_type*, execution_space>;
+  using mat_type = Kokkos::View<scalar_type**, execution_space>;
+
+  static constexpr int neqs = 1;
+
+  TrigonometricEquation() {}
+
+  KOKKOS_FUNCTION void residual(const vec_type& y, const vec_type& f) const {
+    f(0) = Kokkos::cos(y(0)) - y(0);
+  }
+
+  KOKKOS_FUNCTION void jacobian(const vec_type& y, const mat_type& jac) const {
+    jac(0, 0) = -Kokkos::sin(y(0)) - 1;
+  }
+};
+
+// Logarithmic equation
+// f(x) = 7x - log(7x) - 1 = 0
+// Solution: 1/7 = 0.14285714285
+// f'(x) = 7 - (1 / x)
+template <typename execution_space, typename scalar_type>
+struct LogarithmicEquation {
+  using vec_type = Kokkos::View<scalar_type*, execution_space>;
+  using mat_type = Kokkos::View<scalar_type**, execution_space>;
+
+  static constexpr int neqs = 1;
+
+  LogarithmicEquation() {}
+
+  KOKKOS_FUNCTION void residual(const vec_type& y, const vec_type& f) const {
+    f(0) = 7 * y(0) - Kokkos::log(7 * y(0)) - 1;
+  }
+
+  KOKKOS_FUNCTION void jacobian(const vec_type& y, const mat_type& jac) const {
+    jac(0, 0) = 7 - 1 / y(0);
```

Comment:
Also, thanks for the Calc 2 review here. I thought this should be ``7-7/y`` at first but nope :)

Reply:
Yes, `log` is an interesting one... I like this aspect of the ODE work which forces me to review some calculus and functional analysis